r1cs/gadgets/std/src/groups/nonnative/nonnative_group_gadget.rs

```diff
@@ -0,0 +1,844 @@
+        &mut self,
+        mut cs: CS,
+    ) -> Result<(), SynthesisError> {
+        let a = P::COEFF_A;
+
+        let x_squared = self.x.square(cs.ns(|| "x^2"))?;
+
+        let one = P::BaseField::one();
+        let two = one.double();
+        let three = two + &one;
+
+        let three_x_squared = x_squared.mul_by_constant(cs.ns(|| "3 * x^2"), &three)?;
+        let three_x_squared_plus_a = three_x_squared.add_constant(cs.ns(|| "3 * x^2 + a"), &a)?;
+
+        let two_y = self.y.double(cs.ns(|| "2y"))?;
+
+        let lambda = NonNativeFieldGadget::alloc(cs.ns(|| "lambda"), || {
+            let y_doubled_inv = NonNativeFieldGadget::get_value(&two_y).get()?.inverse().get()?;
+            Ok(three_x_squared_plus_a.get_value().get()? * &y_doubled_inv)
+        })?;
+
+        // Check lambda
+        lambda.mul_equals(cs.ns(|| "check lambda"), &two_y, &three_x_squared_plus_a)?;
+
+        // Allocate fresh x and y as a temporary workaround to reduce the R1CS density.
+        let x = NonNativeFieldGadget::alloc(
+            cs.ns(|| "new x"),
+            || {
+                let lambda_val = lambda.get_value().get()?;
+                let x_val = self.x.get_value().get()?;
+                Ok((lambda_val * &lambda_val) - &x_val - &x_val)
+            }
+        )?;
+
+        // lambda * lambda = new_x + 2_old_x
+        let new_x_plus_two_x = self.x
+            .add(cs.ns(|| "2old_x"), &self.x)?
+            .add(cs.ns(|| "new_x + 2old_x"), &x)?;
+        lambda.mul_equals(cs.ns(|| "check new x"), &lambda, &new_x_plus_two_x)?;
+
+        let y = NonNativeFieldGadget::alloc(
+            cs.ns(|| "new y"),
+            || {
+                let lambda_val = lambda.get_value().get()?;
+                let x_val = self.x.get_value().get()?;
+                let y_val = self.y.get_value().get()?;
+                let new_x_val = x.get_value().get()?;
+                Ok(((x_val - &new_x_val) * &lambda_val) - &y_val)
+            }
+        )?;
+
+        //lambda * (old_x - new_x) = new_y + old_y
+        let old_x_minus_new_x = self.x
+            .sub(cs.ns(|| "old_x - new_x"), &x)?;
+        let old_y_plus_new_y = self.y
+            .add(cs.ns(|| "old_y + new_y"), &y)?;
+        lambda.mul_equals(cs.ns(|| "check new y"), &old_x_minus_new_x, &old_y_plus_new_y)?;
+
+        *self = Self::new(x, y, Boolean::constant(false));
+        Ok(())
+    }
+
+
+    fn negate<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+    ) -> Result<Self, SynthesisError> {
+        Ok(Self::new(
+            self.x.clone(),
+            self.y.negate(cs.ns(|| "negate y"))?,
+            self.infinity
+        ))
+    }
+
+
+    /// Incomplete addition: neither `self` nor `other` can be the neutral
+    /// element.
+    fn add_constant<CS: ConstraintSystem<ConstraintF>>(
+        &self,
+        mut cs: CS,
+        other: &SWProjective<P>,
+    ) -> Result<Self, SynthesisError> {
```

Comment:
The same here: add documentation how the circuit treats exceptional cases due to incomplete arithmetics.
Also, why not using the shift method here?